### PR TITLE
Inroduce dimensions in system metrics producer

### PR DESCRIPTION
**Pull request by mairovichaa**

### CODE REVIEW

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractMemoryMetricsProducer.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import java.lang.management.MemoryMXBean;
+import java.lang.management.MemoryPoolMXBean;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractMemoryMetricsProducer extends AbstractMetricsProducer implements MemoryMetricsProducer {
+    public static final MetricName DEFAULT_NAME_PREFIX = MetricName.of("Memory");
+
+    protected static final String INIT_MESSAGE = "The amount of memory in bytes that the Java virtual machine initially requests from the operating system for memory management";
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractMemoryMetricsProducer.java:15]: This name is inconsistent with the DESCRIPTION constants used for the same purpose.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/dimensions/MetricDimensionUtils.java [hunk 2]
```diff
@@ -11,6 +11,10 @@
 @SuppressWarnings("ForLoopReplaceableByForEach")
 public class MetricDimensionUtils {
 
+    public static final MetricDimension NAME_DIMENSION = new MetricDimension("name");
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/dimensions/MetricDimensionUtils.java:14]: This class is supposed to consist of only general-purpose utility methods for working with MetricDimensions. I believe we should place these specific MetricDimensions in the corresponding specific classes in which they are used.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalBufferPoolsMetricsProducer.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package com.ringcentral.platform.metrics.producers.dimensional;
+
+import com.ringcentral.platform.metrics.MetricModBuilder;
+import com.ringcentral.platform.metrics.MetricRegistry;
+import com.ringcentral.platform.metrics.dimensions.MetricDimension;
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalBufferPoolsMetricsProducer.java:5]: Please, remove unused imports.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/SystemMetricsProducer.java [hunk 4]
```diff
@@ -16,7 +18,11 @@ public class SystemMetricsProducer implements MetricsProducer {
     private final ClassesMetricsProducer classesMetricsProducer;
 
     public SystemMetricsProducer() {
-        this(null);
+        this(null, false);
+    }
+
+    public SystemMetricsProducer(boolean isDimensional) {
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/SystemMetricsProducer.java:24]: This is a minor thing and is actually a matter of taste, but the prefix 'is' for boolean fields/params seems redundant.
https://stackoverflow.com/questions/5322648/for-a-boolean-field-what-is-the-naming-convention-for-its-getter-setter

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/SystemMetricsProducer.java [hunk 5]
```diff
@@ -36,19 +43,24 @@ public SystemMetricsProducer(
             threadsMetricNamePrefix,
             bufferPoolsMetricNamePrefix,
             classesMetricNamePrefix,
-            null);
+            null,
+                isDimensional);
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/SystemMetricsProducer.java:47]: I'd suggest using the same indentation for all the parameters including the last one.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/RuntimeMetricsProducer.java [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 import com.ringcentral.platform.metrics.*;
 import com.ringcentral.platform.metrics.names.MetricName;
+import com.ringcentral.platform.metrics.producers.AbstractMetricsProducer;
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/RuntimeMetricsProducer.java:5]: Could you please remove this unused import?

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/ClassesMetricsProducer.java [hunk 7]
```diff
@@ -1,7 +1,9 @@
 package com.ringcentral.platform.metrics.producers;
 
-import com.ringcentral.platform.metrics.*;
+import com.ringcentral.platform.metrics.MetricModBuilder;
+import com.ringcentral.platform.metrics.MetricRegistry;
 import com.ringcentral.platform.metrics.names.MetricName;
+import com.ringcentral.platform.metrics.producers.AbstractMetricsProducer;
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/ClassesMetricsProducer.java:6]: Could you please remove this unused import?

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractMetricsProducer.java [hunk 8]
```diff
@@ -41,22 +44,62 @@ protected MetricName nameWithSuffix(String... suffix) {
...
 
+    protected Supplier<ObjectVarConfigBuilder> objectVarConfigBuilderSupplier(
+            String description, MetricDimension... dimensions
+    ) {
+        final var builder = objectVarConfigBuilder();
+        setDescriptionAndDimensions(builder, description, dimensions);
+        return () -> modified(builder);
+    }
+
     protected Supplier<ObjectVarConfigBuilder> objectVarConfigBuilderSupplier() {
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractMetricsProducer.java:55]: This method is not used anymore. What about removing it? Same for the other xVarConfigBuilderSupplier() methods.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalGarbageCollectorsMetricsProducer.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
+package com.ringcentral.platform.metrics.producers.dimensional;
+
+import com.ringcentral.platform.metrics.MetricModBuilder;
+import com.ringcentral.platform.metrics.MetricRegistry;
+import com.ringcentral.platform.metrics.dimensions.MetricDimension;
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalGarbageCollectorsMetricsProducer.java:5]: Please, remove unused imports.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalMemoryMetricsProducer.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
+package com.ringcentral.platform.metrics.producers.dimensional;
+
+import com.ringcentral.platform.metrics.MetricModBuilder;
+import com.ringcentral.platform.metrics.MetricRegistry;
+import com.ringcentral.platform.metrics.dimensions.MetricDimension;
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalMemoryMetricsProducer.java:5]: Please, remove unused imports.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalOperatingSystemMetricsProducer.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
+package com.ringcentral.platform.metrics.producers.dimensional;
+
+import com.ringcentral.platform.metrics.MetricModBuilder;
+import com.ringcentral.platform.metrics.MetricRegistry;
+import com.ringcentral.platform.metrics.dimensions.MetricDimension;
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalOperatingSystemMetricsProducer.java:5]: Please, remove unused imports.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractOperatingSystemMetricsProducer.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
+package com.ringcentral.platform.metrics.producers;
+
+import com.ringcentral.platform.metrics.MetricModBuilder;
+import com.ringcentral.platform.metrics.MetricRegistry;
+import com.ringcentral.platform.metrics.names.MetricName;
+import com.ringcentral.platform.metrics.var.Var;
```
Comment (devromik) [inline metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractOperatingSystemMetricsProducer.java:6]: Please, remove unused imports.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractMemoryMetricsProducer.java (hunk 1 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractMemoryMetricsProducer.java:15]: Fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalBufferPoolsMetricsProducer.java (hunk 3 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalBufferPoolsMetricsProducer.java:5]: fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/SystemMetricsProducer.java (hunk 4 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/SystemMetricsProducer.java:24]: fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/SystemMetricsProducer.java (hunk 5 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/SystemMetricsProducer.java:47]: fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/ClassesMetricsProducer.java (hunk 7 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/ClassesMetricsProducer.java:6]: Fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractMetricsProducer.java (hunk 8 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractMetricsProducer.java:55]: Removed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractOperatingSystemMetricsProducer.java (hunk 12 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/AbstractOperatingSystemMetricsProducer.java:6]: Fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/RuntimeMetricsProducer.java (hunk 6 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/RuntimeMetricsProducer.java:5]: fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalGarbageCollectorsMetricsProducer.java (hunk 9 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalGarbageCollectorsMetricsProducer.java:5]: fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalOperatingSystemMetricsProducer.java (hunk 11 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/producers/dimensional/DimensionalOperatingSystemMetricsProducer.java:5]: fixed, thank you.

File: metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/dimensions/MetricDimensionUtils.java (hunk 2 above)
Comment (mairovichaa) [reply metrics-facade-base/src/main/java/com/ringcentral/platform/metrics/dimensions/MetricDimensionUtils.java:14]: ok, fixed, thank you.

**Review by devromik — APPROVED** [review]

**Pull request merged**